File: src/main/java/org/micoli/php/symfony/list/RouteService.java
package org.micoli.php.symfony.list;

import com.intellij.openapi.project.Project;
import com.jetbrains.php.lang.psi.elements.PhpAttribute;
import java.util.LinkedHashMap;
import java.util.Map;
import org.jetbrains.annotations.NotNull;
import org.micoli.php.service.attributes.AttributeMapping;
import org.micoli.php.symfony.list.configuration.RoutesConfiguration;

public class RouteService extends AbstractAttributeService<RouteElementDTO, RoutesConfiguration> {
    public RouteService() {
        mapping = new AttributeMapping(new LinkedHashMap<>() {
            {
                put("path", RouteService::getStringableValue);
                put("name", RouteService::getStringableValue);
                put("requirements", RouteService::getStringableValue);
                put("options", RouteService::getStringableValue);
                put("defaults", RouteService::getStringableValue);
                put("host", RouteService::getStringableValue);
                put("methods", RouteService::extractMethods);
            }
        });
    }

    @Override
    protected RouteElementDTO createElementDTO(String className, PhpAttribute attribute, String namespace) {
        Map<String, String> values = mapping.clone().extractValues(attribute);
        return new RouteElementDTO(values.get("path"), values.get("name"), values.get("methods"), className, attribute);
    }

    @Override
    protected String[] getNamespaces() {
        return configuration.namespaces;
    }

    @Override
    protected String getAttributeFQCN() {
        return configuration.attributeFQCN;
    }

    private static @NotNull String extractMethods(PhpAttribute.PhpAttributeArgument attributeArgument) {
        return attributeArgument.getArgument().getValue().replaceAll("[\"'\\[\\]]", "");
    }

    public static RouteService getInstance(Project project) {
        return project.getService(RouteService.class);
    }
}

File: src/main/kotlin/org/micoli/php/symfony/list/RouteService.kt
package org.micoli.php.symfony.list

import com.intellij.openapi.components.Service
import com.intellij.openapi.project.Project
import com.jetbrains.php.lang.psi.elements.PhpAttribute
import com.jetbrains.php.lang.psi.elements.PhpAttribute.PhpAttributeArgument
import java.util.*
import java.util.function.Function
import org.micoli.php.service.attributes.AttributeMapping
import org.micoli.php.symfony.list.configuration.RoutesConfiguration

@Service(Service.Level.PROJECT)
class RouteService(project: Project) :
    AbstractAttributeService<RouteElementDTO, RoutesConfiguration>(project) {
    val mapping =
        AttributeMapping(
            object : LinkedHashMap<String, Function<PhpAttributeArgument, String>>() {
                init {
                    put("path") { getStringableValue(it) }
                    put("name") { getStringableValue(it) }
                    put("requirements") { getStringableValue(it) }
                    put("options") { getStringableValue(it) }
                    put("defaults") { getStringableValue(it) }
                    put("host") { getStringableValue(it) }
                    put("methods") { extractMethods(it) }
                }
            })

    override fun createElementDTO(
        className: String,
        attribute: PhpAttribute,
        namespace: String
    ): RouteElementDTO {
        val values = mapping.clone().extractValues(attribute)
        return RouteElementDTO(
            values["path"] ?: "",
            values["name"] ?: "",
            values["methods"] ?: "",
            className,
            attribute)
    }

    override fun getNamespaces(): Array<String> {
        return configuration!!.namespaces
    }

    override fun getAttributeFQCN(): String {
        return configuration!!.attributeFQCN
    }

    companion object {
        private fun extractMethods(attributeArgument: PhpAttributeArgument): String {
            return attributeArgument.argument.value.replace("[\"'\\[\\]]".toRegex(), "")
        }

        @JvmStatic
        fun getInstance(project: Project): RouteService {
            return project.getService(RouteService::class.java)
        }
    }
}
